File: farebot-app-persist/src/main/java/com/codebutler/farebot/persist/db/DbCardPersister.java
package com.codebutler.farebot.persist.db;

import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.codebutler.farebot.persist.CardPersister;
import com.codebutler.farebot.persist.db.model.SavedCard;
import com.codebutler.farebot.persist.db.model.SavedCardModel;
import com.squareup.sqldelight.SqlDelightStatement;

import java.util.ArrayList;
import java.util.List;

public class DbCardPersister implements CardPersister {

    @NonNull private final FareBotOpenHelper mOpenHelper;

    public DbCardPersister(@NonNull FareBotOpenHelper openHelper) {
        mOpenHelper = openHelper;
    }

    @NonNull
    @Override
    public List<SavedCard> getCards() {
        List<SavedCard> result = new ArrayList<>();
        try (SQLiteDatabase db = mOpenHelper.getReadableDatabase()) {
            SqlDelightStatement query = SavedCard.SELECT_ALL;
            try (Cursor cursor = db.rawQuery(query.statement, query.args)) {
                while (cursor.moveToNext()) {
                    result.add(SavedCard.SELECT_ALL_MAPPER.map(cursor));
                }
            }
        }
        return result;
    }

    @Nullable
    @Override
    public SavedCard getCard(long id) {
        try (SQLiteDatabase db = mOpenHelper.getReadableDatabase()) {
            SqlDelightStatement query = SavedCard.FACTORY.select_by_id(id);
            try (Cursor cursor = db.rawQuery(query.statement, query.args)) {
                if (cursor.moveToFirst()) {
                    return SavedCard.SELECT_ALL_MAPPER.map(cursor);
                }
            }
        }
        return null;
    }

    @Override
    public long insertCard(@NonNull SavedCard savedCard) {
        try (SQLiteDatabase db = mOpenHelper.getWritableDatabase()) {
            SavedCardModel.Insert_row insertRow = new SavedCardModel.Insert_row(db, SavedCard.FACTORY);
            insertRow.bind(
                    savedCard.type(),
                    savedCard.serial(),
                    savedCard.data(),
                    savedCard.scanned_at());
            return insertRow.program.executeInsert();
        }
    }

    public void deleteCard(@NonNull SavedCard savedCard) {
        try (SQLiteDatabase db = mOpenHelper.getWritableDatabase()) {
            SavedCardModel.Delete_by_id deleteRow = new SavedCardModel.Delete_by_id(db);
            deleteRow.bind(savedCard._id());
            deleteRow.program.executeUpdateDelete();
        }
    }
}

File: farebot-app/src/commonMain/kotlin/com/codebutler/farebot/persist/db/DbCardPersister.kt
package com.codebutler.farebot.persist.db

import com.codebutler.farebot.card.CardType
import com.codebutler.farebot.persist.CardPersister
import com.codebutler.farebot.persist.db.model.SavedCard
import kotlin.time.Instant

class DbCardPersister(private val db: FareBotDb) : CardPersister {

    override fun getCards(): List<SavedCard> =
        db.savedCardQueries.selectAll().executeAsList().map { it.toSavedCard() }

    override fun getCard(id: Long): SavedCard? =
        db.savedCardQueries.selectById(id).executeAsOneOrNull()?.toSavedCard()

    override fun insertCard(card: SavedCard): Long {
        db.savedCardQueries.insert(
            type = card.type.name,
            serial = card.serial,
            data_ = card.data,
            scanned_at = card.scannedAt.toEpochMilliseconds()
        )
        return db.savedCardQueries.selectAll().executeAsList().firstOrNull()?.id ?: -1
    }

    override fun deleteCard(card: SavedCard) {
        db.savedCardQueries.deleteById(card.id)
    }
}

private fun Cards.toSavedCard() = SavedCard(
    id = id,
    type = CardType.valueOf(type),
    serial = serial,
    data = data_,
    scannedAt = Instant.fromEpochMilliseconds(scanned_at)
)
